File: bin/day-9.py
"""
The classic travelling salesman.  distances are only given in one direction,
but it is assumed you can go in either direction.

Since there are only 8 destinations, the brute force solution is the easiest way
and the builtin itertools.permutations takes all the work out of it.
all that's left is summing up the distances and finding the best (and worst for part 2)

"""
from collections import defaultdict

import itertools
import sys
import time


distances = defaultdict(dict)

for line in open("../inputs/day-9.txt"):
    l, r = line.strip().split(" = ")
    c1, c2 = l.split(" to ")
    distances[c1][c2] = int(r)
    distances[c2][c1] = int(r)

best = sys.maxint
worst = 0

now = time.time()
for perm in itertools.permutations(distances.keys(), len(distances)):
    dist = 0
    start = perm[0]
    for city in perm[1:]:
        dist += distances[start][city]
        start = city
    if dist < best:
        best = dist
    if dist > worst:
        worst = dist

print best
print worst
print (time.time() - now) * 1000

File: bin/day-9.dart
import "dart:io";


Iterable<List> permutations(Iterable src) sync* {
  if (src.length == 1) {
    yield src;
  } else {
    for (var item in src) {
      for (List p in permutations(src.where((i) => i != item).toList())) {
        yield p..add(item);
      }
    }
  }
}


void main() {
  Set visited = new Set<String>();
  int totalDistance = 0;
  Map<String, Map<String, int>> distances = new Map<String, Map<String,int>>();

  List<String> lines = new File("../inputs/day-9.txt").readAsLinesSync();
  Stopwatch watch = new Stopwatch()..start();
  lines.forEach((line) {
    var parts = line.split(" = ");
    var parts2 = parts[0].split(" to ");
    var c1 = parts2[0];
    var c2 = parts2[1];
    int dist = int.parse(parts[1]);
    if (!distances.containsKey(c1)) {
      distances[c1] = new Map<String, int>();
    }
    if (!distances.containsKey(c2)) {
      distances[c2] = new Map<String, int>();
    }
    distances[c1][c2] = dist;
    distances[c2][c1] = dist;
  });

  int worst = 0;
  int best = 9999999;

  for (List<String> perm in permutations(distances.keys)) {
    int dist = 0;
    String start = perm[0];
    for (String city in perm.skip(1)) {
      dist += distances[start][city];
      start = city;
    }
    if (dist > worst) {
      worst = dist;
    }
    if (dist < best) {
      best = dist;
    }
  }

  watch.stop();
  print("Best: $best");
  print("Worst: $worst");
  print("Time: ${watch.elapsedMilliseconds}");
}